feat(linter): import/no-useless-path-segments

rtsao opened the pull request:
Implement the `import/no-useless-path-segments` rule (part of https://github.com/oxc-project/oxc/issues/1117): 

- [ESLint rule docs](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-useless-path-segments.md)
- [ESLint rule source](https://github.com/import-js/eslint-plugin-import/blob/main/src/rules/no-useless-path-segments.js)
- [ESLint rule tests](https://github.com/import-js/eslint-plugin-import/blob/main/tests/src/rules/no-useless-path-segments.js)

I ported the original JS test cases over 1:1 and tried to follow the original JS logic as closely as possible. 

Note that this implementation avoids additional filesystem lookups where possible (i.e. ESM) but needs to for CommonJS (not in `module_record`).

rtsao committed:
Implement import/no-useless-path-segments

camc314 committed:
fix clippy

camc314 reviewed (changes requested):
Thanks for working on this. Unfortunately the CI is failing.

rtsao committed:
Remove invalid code fence

rtsao committed:
Update codegen

rtsao committed:
Remove code fences

rtsao committed:
Address feedback

Boshen commented:
> ## [CodSpeed Performance Report](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?utm_source=github&utm_medium=comment&utm_content=header)
> ### Merging #14569 will **degrade performances by 5.86%**
> Comparing `rtsao:no-useless-path-segments` (625ea87) with `main` ([ba76081](https://github.com/oxc-project/oxc/commit/ba76081ff7c80d4a95909aafd62a46fa3f54077b))[1](#user-content-fn-unexpected-base-4f10f472852554495938db24edb676ac)
> 
> ### Summary
> `❌ 3` regressions `✅ 1` untouched `⏩ 33` skipped[2](#user-content-fn-skipped-4f10f472852554495938db24edb676ac)
> 
> > ⚠️ _Please fix the performance issues or [acknowledge them on CodSpeed](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?utm_source=github&utm_medium=comment&utm_content=acknowledge)._
> 
> ### Benchmarks breakdown
> Mode	Benchmark	`BASE`	`HEAD`	Change
> ❌	Simulation	[`linter[binder.ts]`](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?uri=tasks%2Fbenchmark%2Fbenches%2Flinter.rs%3A%3Alinter%3A%3Abench_linter%3A%3Alinter%5Bbinder.ts%5D&runnerMode=Instrumentation&utm_source=github&utm_medium=comment&utm_content=benchmark)	46.7 ms	49.6 ms	-5.86%
> ❌	Simulation	[`linter[cal.com.tsx]`](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?uri=tasks%2Fbenchmark%2Fbenches%2Flinter.rs%3A%3Alinter%3A%3Abench_linter%3A%3Alinter%5Bcal.com.tsx%5D&runnerMode=Instrumentation&utm_source=github&utm_medium=comment&utm_content=benchmark)	465.4 ms	479.8 ms	-3.02%
> ❌	Simulation	[`linter[react.development.js]`](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?uri=tasks%2Fbenchmark%2Fbenches%2Flinter.rs%3A%3Alinter%3A%3Abench_linter%3A%3Alinter%5Breact.development.js%5D&runnerMode=Instrumentation&utm_source=github&utm_medium=comment&utm_content=benchmark)	18.7 ms	19.4 ms	-3.46%
> ## Footnotes
> 1. No successful run was found on `main` ([9de1322](https://github.com/oxc-project/oxc/commit/9de1322765d22a8fe5bc7d6f6921a99e08425fe8)) during the generation of this report, so [ba76081](https://github.com/oxc-project/oxc/commit/ba76081ff7c80d4a95909aafd62a46fa3f54077b) was used instead as the comparison base. There might be some changes unrelated to this pull request in this report. [↩](#user-content-fnref-unexpected-base-4f10f472852554495938db24edb676ac)
> 2. 33 benchmarks were skipped, so the baseline results were used instead. If they were deleted from the codebase, [click here and archive them to remove them from the performance reports](https://codspeed.io/oxc-project/oxc/branches/rtsao%3Ano-useless-path-segments?sectionId=benchmark-comparison-section-baseline-result-skipped&utm_source=github&utm_medium=comment&utm_content=archive). [↩](#user-content-fnref-skipped-4f10f472852554495938db24edb676ac)

Performance regression is not ideal, probably due to calling `canonicalize`.

camchenry commented:
Thanks for the contribution, I am going to close this for now.

rtsao commented:
@camchenry I think the slow part of this rule is CommonJS support.

If performance is the primary concern, do you think it would be acceptable for this rule to apply only to ESM imports?

I've updated and rebased my branch to this simpler implementation that omits CJS support: https://github.com/oxc-project/oxc/compare/main...rtsao:oxc:no-useless-path-segments

nunomarks commented:
Are we pushing this forward at all ?

rtsao commented:
@nunomarks I'd love to get this rule implemented but am still looking for clarity on whether eslint parity for CommonJS is important or not (which is the slow part of this rule).

@camchenry would you prefer I open a fresh PR? Or if you re-open this one I can push my faster ESM-only implementation
